src/server/json_family.cc

```diff
@@ -554,32 +438,75 @@ size_t CountJsonFields(const JsonType& j) {
   return res;
 }
 
-void SendJsonValue(RedisReplyBuilder* rb, const JsonType& j) {
-  if (j.is_double()) {
-    rb->SendDouble(j.as_double());
-  } else if (j.is_number()) {
-    rb->SendLong(j.as_integer<long>());
-  } else if (j.is_bool()) {
-    rb->SendSimpleString(j.as_bool() ? "true" : "false");
-  } else if (j.is_null()) {
-    rb->SendNull();
-  } else if (j.is_string()) {
-    rb->SendBulkString(j.as_string_view());
-  } else if (j.is_object()) {
-    rb->StartArray(j.size() + 1);
-    rb->SendSimpleString("{");
-    for (const auto& item : j.object_range()) {
-      rb->StartArray(2);
-      rb->SendBulkString(item.key());
-      SendJsonValue(rb, item.value());
-    }
-  } else if (j.is_array()) {
-    rb->StartArray(j.size() + 1);
-    rb->SendSimpleString("[");
-    for (const auto& item : j.array_range()) {
-      SendJsonValue(rb, item);
+template <typename T> struct is_optional : std::false_type {};
+
+template <typename T> struct is_optional<std::optional<T>> : std::true_type {};
+
+template <typename T>
+typename std::enable_if_t<!is_optional<T>::value, OpResult<JsonCallbackResult<T>>>
+ReturnWrongTypeOnNullOpt(JsonCallbackResult<T> result) {
+  return result;
+}
+
+template <typename T>
+typename std::enable_if_t<is_optional<T>::value, OpResult<JsonCallbackResult<T>>>
+ReturnWrongTypeOnNullOpt(JsonCallbackResult<T> result) {
+  if (result.IsV1()) {
+    auto& as_v1 = result.AsV1();
+    if (!as_v1 || !as_v1.value()) {
+      return OpStatus::WRONG_JSON_TYPE;
     }
   }
+  return result;
+}
```

Comment:
Heavy artillery, once you have is_optional you can use if constexpr again 🙂

Reply:
yes, I've thought about it. But this code is repeated twice in `JsonEvaluateOperation` and `UpdateEntry`, so why not to implement it as a function?